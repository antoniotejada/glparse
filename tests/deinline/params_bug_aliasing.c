// Test for the following unfixed pointer aliasing bug:
// Multiple calls to
//   openAsset(param_AAssetManager_ptr_0, "filename", &local_AAsset_ptr_0)
//   getAssetBuffer(local_AAsset_ptr_0);
// are deinlined as
//   openAsset(param_AAssetManager_ptr_0, param_char_ptr_1, param_AAsset_ptr_ptr_2);
//   getAssetBuffer(param_AAsset_ptr_3);
// Note how getAssetBuffer calls using a temporary variable that is no
// longer updated by the call to openAsset
// This is currently fixed in the generator by unifying those calls
// into a single openAndGetAssetBuffer
// A better fix would be to do parameter coalescing in an aliasing-aware way

void a(AAssetManager* param_AAssetManager_ptr_0)
{
    AAsset* local_AAsset_ptr_0 = NULL;

    openAsset(param_AAssetManager_ptr_0, "filename", &local_AAsset_ptr_0)
    getAssetBuffer(local_AAsset_ptr_0);
    openAsset(param_AAssetManager_ptr_0, "filename1", &local_AAsset_ptr_0)
    getAssetBuffer(local_AAsset_ptr_0);
    openAsset(param_AAssetManager_ptr_0, "filename2", &local_AAsset_ptr_0)
    getAssetBuffer(local_AAsset_ptr_0);
}

void f(AAssetManager* param_AAssetManager_ptr_0)
{
    a(param_AAssetManager_ptr_0);
    a(param_AAssetManager_ptr_0);
    a(param_AAssetManager_ptr_0);
    a(param_AAssetManager_ptr_0);
    a(param_AAssetManager_ptr_0);
    a(param_AAssetManager_ptr_0);
}