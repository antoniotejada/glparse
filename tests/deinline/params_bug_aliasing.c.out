// Src lines 13
// Dest lines 11
// Test for the following unfixed pointer aliasing bug:
// Multiple calls to
//   openAsset(param_AAssetManager_ptr_0, "filename", &local_AAsset_ptr_0)
//   getAssetBuffer(local_AAsset_ptr_0);
// are deinlined as
//   openAsset(param_AAssetManager_ptr_0, param_char_ptr_1, param_AAsset_ptr_ptr_2);
//   getAssetBuffer(param_AAsset_ptr_3);
// Note how getAssetBuffer calls using a temporary variable that is no
// longer updated by the call to openAsset
// This is currently fixed in the generator by unifying those calls
// into a single openAndGetAssetBuffer
// A better fix would be to do parameter coalescing in an aliasing-aware way


void a(AAssetManager* param_AAssetManager_ptr_0);
void f(AAssetManager* param_AAssetManager_ptr_0);
void subframe2(AAssetManager * param_AAssetManager_ptr_0, char * param_char_ptr_1, AAsset * * param_AAsset_ptr_ptr_2, AAsset * param_AAsset_ptr_3);
void subframe3(AAssetManager * param_AAssetManager_ptr_0);

void a(AAssetManager* param_AAssetManager_ptr_0)
{
    AAsset* local_AAsset_ptr_0 = NULL;
    
    subframe2(param_AAssetManager_ptr_0, "filename", &local_AAsset_ptr_0, local_AAsset_ptr_0);
    subframe2(param_AAssetManager_ptr_0, "filename1", &local_AAsset_ptr_0, local_AAsset_ptr_0);
    subframe2(param_AAssetManager_ptr_0, "filename2", &local_AAsset_ptr_0, local_AAsset_ptr_0);
}

void f(AAssetManager* param_AAssetManager_ptr_0)
{
    subframe3(param_AAssetManager_ptr_0);
    subframe3(param_AAssetManager_ptr_0);
    subframe3(param_AAssetManager_ptr_0);
}

void subframe2(AAssetManager * param_AAssetManager_ptr_0, char * param_char_ptr_1, AAsset * * param_AAsset_ptr_ptr_2, AAsset * param_AAsset_ptr_3)
{
    openAsset(param_AAssetManager_ptr_0, param_char_ptr_1, param_AAsset_ptr_ptr_2);
    getAssetBuffer(param_AAsset_ptr_3);
}

void subframe3(AAssetManager * param_AAssetManager_ptr_0)
{
    a(param_AAssetManager_ptr_0);
    a(param_AAssetManager_ptr_0);
}
